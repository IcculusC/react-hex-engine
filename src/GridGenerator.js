import Hex from "./models/Hex";
import HexUtils from "./HexUtils";

class GridGenerator {
  static getGenerator(name) {
    if (GridGenerator.hasOwnProperty(name)) return GridGenerator[name];

    return null;
  }

  static ring(center, mapRadius) {
    let hexas = [];
    let hex = HexUtils.add(
      center,
      HexUtils.multiply(HexUtils.direction(4), mapRadius)
    );
    for (let i = 0; i < 6; i++) {
      for (let j = 0; j < mapRadius; j++) {
        hexas.push(hex);
        hex = HexUtils.neighbour(hex, i);
      }
    }
    return hexas;
  }

  static spiral(center, mapRadius) {
    let results = [center];
    for (let k = 1; k <= mapRadius; k++) {
      const temp = this.ring(center, k);
      results = results.concat(temp);
    }
    return results;
  }

  static parallelogram(q1, q2, r1, r2) {
    let hexas = [];
    for (let q = q1; q <= q2; q++) {
      for (let r = r1; r <= r2; r++) {
        hexas.push(new Hex(q, r, -q - r));
      }
    }

    return hexas;
  }

  static triangle(mapSize) {
    let hexas = [];
    for (let q = 0; q <= mapSize; q++) {
      for (let r = 0; r <= mapSize - q; r++) {
        hexas.push(new Hex(q, r, -q - r));
      }
    }

    return hexas;
  }

  static hexagon(mapRadius) {
    let hexas = [];
    for (let q = -mapRadius; q <= mapRadius; q++) {
      let r1 = Math.max(-mapRadius, -q - mapRadius);
      let r2 = Math.min(mapRadius, -q + mapRadius);
      for (let r = r1; r <= r2; r++) {
        hexas.push(new Hex(q, r, -q - r));
      }
    }

    return hexas;
  }

  static rectangle(mapWidth, mapHeight) {
    let hexas = [];
    for (let r = 0; r < mapHeight; r++) {
      let offset = Math.floor(r / 2); // or r>>1
      for (let q = -offset; q < mapWidth - offset; q++) {
        hexas.push(new Hex(q, r, -q - r));
      }
    }

    return hexas;
  }

  static orientedRectangle(mapWidth, mapHeight) {
    let hexas = [];
    for (let q = 0; q < mapWidth; q++) {
      let offset = Math.floor(q / 2); // or q>>1
      for (let r = -offset; r < mapHeight - offset; r++) {
        hexas.push(new Hex(q, r, -q - r));
      }
    }

    return hexas;
  }
}

export default GridGenerator;
