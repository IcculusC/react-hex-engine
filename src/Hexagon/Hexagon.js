import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Hex from "../models/Hex";
import HexUtils from "../HexUtils";
import Point from "../models/Point";
import { withExpandedLayout } from "../Context";

class Hexagon extends Component {
  static propTypes = {
    children: PropTypes.node,
    classes: PropTypes.objectOf(PropTypes.any),
    data: PropTypes.object,
    highlighted: PropTypes.bool,
    layout: PropTypes.objectOf(PropTypes.any).isRequired,
    onClick: PropTypes.func,
    onDragEnd: PropTypes.func,
    onDragOver: PropTypes.func,
    onDragStart: PropTypes.func,
    onDrop: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onMouseOver: PropTypes.func,
    points: PropTypes.string.isRequired,
    q: PropTypes.number.isRequired,
    r: PropTypes.number.isRequired,
    s: PropTypes.number.isRequired,
    selected: PropTypes.bool,
    showCoordinates: PropTypes.bool
  };

  static defaultProps = {
    className: "",
    classes: {
      group: "",
      hexagon: "",
      highlighted: "",
      hovered: "",
      polygon: "",
      q: "",
      r: "",
      s: "",
      selected: ""
    },
    showCoordinates: false
  };

  static getCoordinateTextOffset(
    corner,
    { orientation, size },
    offset = new Point(0, 0),
    scale = new Point(0.75, 0.75)
  ) {
    const angle = (2.0 * Math.PI * (corner + 0.5)) / 6;
    return new Point(
      size.x * scale.x * Math.cos(angle),
      size.y * scale.y * Math.sin(angle)
    );
  }

  state = {
    hex: {},
    hovered: false
  };

  constructor(props) {
    super(props);
    const hex = new Hex(props.q, props.r, props.s);
    const pixel = HexUtils.hexToPixel(hex, props.layout);
    this.state = {
      ...this.state,
      hex,
      pixel
    };
  }

  onClick(e) {
    if (this.props.onClick) {
      this.props.onClick(e, this.state.hex);
    }
  }

  onDragEnd(e) {
    if (this.props.onDragEnd) {
      e.preventDefault();
      const success = e.dataTransfer.dropEffect !== "none";
      this.props.onDragEnd(e, this.state.hex, success);
    }
  }

  onDragOver(e) {
    if (this.props.onDragOver) {
      this.props.onDragOver(e, this.state.hex);
    }
  }

  onDragStart(e) {
    if (this.props.onDragStart) {
      const targetProps = {
        ...this.state,
        data: this.props.data,
        classes: this.props.classes
      };
      e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
      this.props.onDragStart(e, this.state.hex);
    }
  }

  onDrop(e) {
    if (this.props.onDrop) {
      e.preventDefault();
      const target = JSON.parse(e.dataTransfer.getData("hexagon"));
      this.props.onDrop(e, this.state.hex, target);
    }
  }

  onMouseEnter(e) {
    this.setState({ hovered: true });
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e, this.state.hex);
    }
  }

  onMouseLeave(e) {
    this.setState({ hovered: false });
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e, this.state.hex);
    }
  }

  onMouseOver(e) {
    if (this.props.onMouseOver) {
      this.props.onMouseOver(e, this.state.hex);
    }
  }

  render() {
    const {
      classes,
      highlighted,
      layout,
      points,
      q,
      r,
      s,
      selected,
      showCoordinates
    } = this.props;
    const { hex, hovered, pixel } = this.state;
    let qPixel, rPixel, sPixel;
    if (showCoordinates) {
      qPixel = Hexagon.getCoordinateTextOffset(3, layout, {
        x: 0,
        y: 1
      });
      rPixel = Hexagon.getCoordinateTextOffset(1, layout, {
        x: -1,
        y: -1
      });
      sPixel = Hexagon.getCoordinateTextOffset(5, layout, {
        x: -2,
        y: 1
      });
    }
    return (
      <g
        className={classNames("hexagon-group", classes.group)}
        draggable="true"
        onClick={e => this.onClick(e)}
        onDragEnd={e => this.onDragEnd(e)}
        onDragOver={e => this.onDragOver(e)}
        onDragStart={e => this.onDragStart(e)}
        onDrop={e => this.onDrop(e)}
        onMouseEnter={e => this.onMouseEnter(e)}
        onMouseLeave={e => this.onMouseLeave(e)}
        onMouseOver={e => this.onMouseOver(e)}
        transform={`translate(${pixel.x}, ${pixel.y})`}
      >
        <g
          className={classNames("hexagon", classes.hexagon, {
            [classes.selected]: selected,
            [classes.highlighted]: highlighted,
            [classes.hovered]: hovered
          })}
        >
          <polygon className={classes.polygon} points={points} />
          {this.props.children}
          {showCoordinates && (
            <React.Fragment>
              <text
                {...qPixel}
                className={classes.q}
                fontSize={2}
                textAnchor="middle"
              >
                {q}
              </text>
              <text
                {...rPixel}
                className={classes.r}
                fontSize={2}
                textAnchor="middle"
              >
                {r}
              </text>
              <text
                {...sPixel}
                className={classes.s}
                fontSize={2}
                textAnchor="middle"
              >
                {s}
              </text>
            </React.Fragment>
          )}
        </g>
      </g>
    );
  }
}

export default withExpandedLayout(Hexagon);
