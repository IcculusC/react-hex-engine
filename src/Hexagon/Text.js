import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

class Text extends Component {
  static propTypes = {
    anchor: PropTypes.string,
    children: PropTypes.string,
    className: PropTypes.string,
    classes: PropTypes.objectOf(PropTypes.string),
    x: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    y: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };

  static defaultProps = {
    anchor: "middle",
    classes: { text: "" }
  };

  render() {
    const { anchor, children, className, classes, x, y } = this.props;
    return (
      <text
        className={classNames(className, classes.text)}
        textAnchor={anchor}
        x={x || 0}
        y={y ? y : "0.3em"}
      >
        {children}
      </text>
    );
  }
}

export default Text;
