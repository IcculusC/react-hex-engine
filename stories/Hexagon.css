.hexagon {
    fill: #00a;
    transition: fill 0.2s;
}

.hexagon.hovered {
    fill: #aa0;
    transition: fill 0.2s;
}

.hexagon.hovered > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.selected {
    fill: #ff0;
    transition: fill 0.5s;
}

.hexagon.selected > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.selected.hovered {
    fill: #fff;
}

.hexagon.selected.hovered > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.highlighted {
    fill: blue;
}

.hexagon.showCoordinates > text.axis {
    fill: #fff;
    font-size: 2px;
    font-family: sans-serif;
}

.hexagon.text > text.text {
    fill: #fff;
    font-size: 3px;
    font-family: sans-serif;
}
