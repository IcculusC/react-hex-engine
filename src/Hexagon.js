import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Hex from "./models/Hex";
import HexUtils from "./HexUtils";
import Point from "./models/Point";
import Text from "./Text";
import { withHexEngine } from "./Context";

class Hexagon extends Component {
  static propTypes = {
    children: PropTypes.node,
    classes: PropTypes.objectOf(PropTypes.any),
    data: PropTypes.object,
    hoverable: PropTypes.bool,
    /** Provides a means to highlight a tile without selecting it, styleable with the `highlighted` class */
    highlighted: PropTypes.bool,
    layout: PropTypes.objectOf(PropTypes.any).isRequired,
    onClick: PropTypes.func,
    onDragEnd: PropTypes.func,
    onDragOver: PropTypes.func,
    onDragStart: PropTypes.func,
    onDrop: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onMouseOver: PropTypes.func,
    points: PropTypes.string.isRequired,
    q: PropTypes.number.isRequired,
    r: PropTypes.number.isRequired,
    s: PropTypes.number.isRequired,
    /** Disable the `selected` prop, rather redundant */
    selectable: PropTypes.bool,
    selected: PropTypes.bool,
    /** Show the `q`, `r`, `s` coordinates of the hexagon, styleable with the `q`, `r`, and `s` classes */
    showCoordinates: PropTypes.bool,
    /** Show text in the hexagon, styleable with the `text` class */
    text: PropTypes.string,
    TextProps: PropTypes.objectOf(PropTypes.any)
  };

  static defaultProps = {
    classes: {
      group: "",
      hexagon: "",
      highlighted: "",
      hovered: "",
      polygon: "",
      q: "",
      r: "",
      s: "",
      selected: "",
      text: ""
    },
    data: {},
    highlighted: false,
    hoverable: true,
    selectable: true,
    selected: false,
    showCoordinates: false,
    text: "",
    TextProps: {}
  };

  static getCoordinateTextOffset(
    corner,
    { orientation, size },
    offset = new Point(0, 0),
    scale = new Point(0.75, 0.75)
  ) {
    const angle = (2.0 * Math.PI * (corner + 0.5)) / 6;
    return new Point(
      size.x * scale.x * Math.cos(angle),
      size.y * scale.y * Math.sin(angle)
    );
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.points !== nextProps.points) {
      const { layout, points, q, r } = nextProps;
      const hex = new Hex(q, r);
      const pixel = HexUtils.hexToPixel(hex, nextProps.layout);
      return { hex, pixel, layout, points };
    }
    return null;
  }

  state = {
    hex: {},
    hovered: false,
    layout: {},
    pixel: {},
    points: ""
  };

  constructor(props) {
    super(props);
    const hex = new Hex(props.q, props.r);
    const pixel = HexUtils.hexToPixel(hex, props.layout);
    this.state = {
      ...this.state,
      hex,
      pixel
    };
  }

  onClick(e) {
    if (this.props.onClick) {
      this.props.onClick(e, this.state.hex);
    }
  }

  onDragEnd(e) {
    if (this.props.onDragEnd) {
      e.preventDefault();
      const success = e.dataTransfer.dropEffect !== "none";
      this.props.onDragEnd(e, this.state.hex, success);
    }
  }

  onDragOver(e) {
    if (this.props.onDragOver) {
      this.props.onDragOver(e, this.state.hex);
    }
  }

  onDragStart(e) {
    if (this.props.onDragStart) {
      const targetProps = {
        ...this.state,
        data: this.props.data,
        classes: this.props.classes
      };
      e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
      this.props.onDragStart(e, this.state.hex);
    }
  }

  onDrop(e) {
    if (this.props.onDrop) {
      e.preventDefault();
      const target = JSON.parse(e.dataTransfer.getData("hexagon"));
      this.props.onDrop(e, this.state.hex, target);
    }
  }

  onMouseEnter(e) {
    this.setState({ hovered: true });
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e, this.state.hex);
    }
  }

  onMouseLeave(e) {
    this.setState({ hovered: false });
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e, this.state.hex);
    }
  }

  onMouseOver(e) {
    if (this.props.onMouseOver) {
      this.props.onMouseOver(e, this.state.hex);
    }
  }

  render() {
    const {
      classes,
      highlighted,
      hoverable,
      layout,
      points,
      q,
      r,
      s,
      selectable,
      selected,
      showCoordinates,
      text,
      TextProps
    } = this.props;
    const { hovered, pixel } = this.state;
    let qPixel, rPixel, sPixel;
    if (showCoordinates) {
      qPixel = Hexagon.getCoordinateTextOffset(3, layout, {
        x: 0,
        y: 1
      });
      rPixel = Hexagon.getCoordinateTextOffset(1, layout, {
        x: -1,
        y: -1
      });
      sPixel = Hexagon.getCoordinateTextOffset(5, layout, {
        x: -2,
        y: 1
      });
    }
    return (
      <g
        className={classNames("hexagon-group", classes.group)}
        draggable="true"
        onClick={e => this.onClick(e)}
        onDragEnd={e => this.onDragEnd(e)}
        onDragOver={e => this.onDragOver(e)}
        onDragStart={e => this.onDragStart(e)}
        onDrop={e => this.onDrop(e)}
        onMouseEnter={e => this.onMouseEnter(e)}
        onMouseLeave={e => this.onMouseLeave(e)}
        onMouseOver={e => this.onMouseOver(e)}
        transform={`translate(${pixel.x}, ${pixel.y})`}
      >
        <g
          className={classNames("hexagon", classes.hexagon, {
            [classes.selected]: selectable && selected,
            [classes.highlighted]: highlighted,
            [classes.hovered]: hoverable && hovered
          })}
        >
          <polygon className={classes.polygon} points={points} />
          {this.props.children}
          {text ? (
            <Text classes={{ text: classes.text }} {...TextProps}>
              {text}
            </Text>
          ) : null}
          {showCoordinates ? (
            <React.Fragment>
              <text
                {...qPixel}
                className={classes.q}
                fontSize={2}
                textAnchor="middle"
              >
                {q}
              </text>
              <text
                {...rPixel}
                className={classes.r}
                fontSize={2}
                textAnchor="middle"
              >
                {r}
              </text>
              <text
                {...sPixel}
                className={classes.s}
                fontSize={2}
                textAnchor="middle"
              >
                {s}
              </text>
            </React.Fragment>
          ) : null}
        </g>
      </g>
    );
  }
}

export const Hexagon_ = Hexagon;
export default withHexEngine(Hexagon_);
