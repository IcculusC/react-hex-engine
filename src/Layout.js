import React, { Component } from 'react';
import PropTypes from 'prop-types';
import HexUtils from './HexUtils';
import Orientation from './models/Orientation';
import Point from './models/Point';
import Hexagon from './Hexagon/Hexagon';
import Path from './Path';
import { ViewBoxConsumer } from './HexGrid';

export const LAYOUT_FLAT = new Orientation(3.0 / 2.0, 0.0, Math.sqrt(3.0) / 2.0, Math.sqrt(3.0),2.0 / 3.0, 0.0, -1.0 / 3.0, Math.sqrt(3.0) / 3.0, 0.0);
export const LAYOUT_POINTY = new Orientation(Math.sqrt(3.0), Math.sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0, Math.sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0, 0.5);

class Layout extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    flat: PropTypes.bool,
    origin: PropTypes.object,
    size: PropTypes.object,
    spacing: PropTypes.number
  };

  static defaultProps = {
    size: new Point(10, 10),
    flat: true,
    spacing: 1.0,
    origin: new Point(0, 0)
  }

  getPointOffset(corner, orientation, size) {
    let angle = 2.0 * Math.PI * (corner + orientation.startAngle) / 6;
    return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle));
  }

  // TODO Refactor
  calculateCoordinates(orientation) {
    const corners = [];
    const center = new Point(0, 0);
    const { size } = this.props;

    Array.from(new Array(6), (x, i) => {
      const offset = this.getPointOffset(i, orientation, size);
      const point = new Point(center.x + offset.x, center.y + offset.y);
      corners.push(point);
    });

    return corners;
  }

  filterChildren(children, cornerCoords, viewBox, layout) {
    const { x, y, width, height } = viewBox;
    const filtered_ = React.Children.toArray(children).filter(child => {
      if (!child.props) {
        return true;
      }
      if (child.type === Hexagon || (child.props.q !== undefined && child.props.r !== undefined && child.props.s !== undefined)) {
        const point = HexUtils.hexToPixel(child.props, layout);
        const corners = cornerCoords.map(coord => new Point(coord.x + point.x, coord.y + point.y));
        const filtered = corners.filter(corner => corner.x > x && corner.x < width + x && corner.y > y && corner.y < + height + y);
        return filtered.length;
      }
      return true;
    });

    return filtered_;
  }

  render() {
    const { children = [], flat, className, viewBox, ...rest } = this.props;
    const orientation = (flat) ? LAYOUT_FLAT : LAYOUT_POINTY;
    const cornerCoords = this.calculateCoordinates(orientation);
    const points = cornerCoords.map(point => `${point.x},${point.y}`).join(' ');
    const layout = {...rest, orientation};
    const inBounds = this.filterChildren(children, cornerCoords, viewBox, layout);
    return (
      <LayoutProvider value={{ layout, points }}>
        <g className={className}>
          {inBounds}
        </g>
      </LayoutProvider>
    );
  }
}

export const LayoutContext = React.createContext({
  layout: LAYOUT_FLAT,
  points: ""
})
export const { Provider: LayoutProvider, Consumer: LayoutConsumer } = LayoutContext;

export default props => <ViewBoxConsumer>{viewBox => <Layout {...props} viewBox={viewBox} />}</ViewBoxConsumer>;

