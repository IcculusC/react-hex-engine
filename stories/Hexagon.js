import React from "react";
import { storiesOf } from "@storybook/react";
import addons from "@storybook/addons";
import Events from "@storybook/core-events";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import { withInfo } from "@storybook/addon-info";
import Hexagon, { Hexagon_ } from "../src/Hexagon";
import HexEngine from "../src/HexEngine";
import Point from "../src/models/Point";
import { manager } from "@storybook/addon-knobs/dist/registerKnobs";
import "./Hexagon.css";

Hexagon.displayName = "Hexagon";
Hexagon_.displayName = "Hexagon";
HexEngine.displayName = "HexEngine";

const stories = storiesOf("Hexagon", module);

stories.addDecorator(withKnobs);
stories.addDecorator(withInfo);

stories.add(
  "Kitchen Sink",
  () => {
    const selectable = boolean("selectable", true, "Hexagon");
    const selected = boolean("selected", false, "Hexagon");
    const highlighted = boolean("highlighted", false, "Hexagon");
    const hoverable = boolean("hoverable", true, "Hexagon");
    const showCoordinates = boolean("showCoordinates", false, "Hexagon");
    const innerText = text("text", "", "Hexagon");

    return (
      <HexEngine
        size={new Point(10, 10)}
        width={320}
        height={240}
        viewBox={{
          x: -25,
          y: -25,
          width: 50,
          height: 50
        }}
      >
        <Hexagon
          q={0}
          r={0}
          s={0}
          selectable={selectable}
          selected={selected}
          highlighted={highlighted}
          hoverable={hoverable}
          text={innerText}
          showCoordinates={showCoordinates}
          classes={{
            hexagon: "showCoordinates text",
            highlighted: "highlighted",
            hovered: "hovered",
            selected: "selected",
            q: "axis",
            r: "axis",
            s: "axis",
            text: "text"
          }}
          onClick={() => {
            const { knobStore } = manager;
            const knobOptions = knobStore.get("selected");
            knobOptions.value = !selected;
            knobStore.markAllUnused();
            addons.getChannel().emit(Events.FORCE_RE_RENDER);
          }}
        />
      </HexEngine>
    );
  },
  {
    info: {
      header: false,
      inline: true,
      maxPropStringLength: 512,
      propTables: [Hexagon_], // should be able to fix this with useContext in react 16.7.x
      propTablesExclude: [HexEngine, Hexagon],
      source: false,
      excludedPropTypes: [
        "data",
        "layout",
        "onDragEnd",
        "onDragOver",
        "onDragStart",
        "onDragStop",
        "onDrop",
        "points"
      ]
    }
  }
);
