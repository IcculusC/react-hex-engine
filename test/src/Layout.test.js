import React from 'react';
import renderer from 'react-test-renderer';

import Layout from '../../src/Layout';
import Hexagon from '../../src/Hexagon/Hexagon';
import GridGenerator from '../../src/GridGenerator';

test('Layout should render correctly with default props', () => {
  const tree = renderer.create(
    <Layout
      className={'test1'}
    >
      <div>child</div>
    </Layout>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Layout should render correctly with custom props', () => {
  const tree = renderer.create(
    <Layout
      className={'test2'}
      flat={false}
      origin={{ x: 2, y: 4 }}
      size={{ x: 12, y: 14 }}
      spacing={2.0}
    >
      <div>child</div>
    </Layout>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Layout should only render hexagons in bounds', () => {
  const tree = renderer.create(
    <Layout
      viewBox={{ x: -50, y: -50, width: 100, height: 100 }}
      className={'test2'}
      flat={false}
      origin={{ x: 0, y: 0 }}
      size={{ x: 12, y: 14 }}
      spacing={2.0}
    >
      <Hexagon q={0} r={0} s={0} />
      <Hexagon q={500} r={500} s={500} />
    </Layout>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Layout should only render hexagons in bounds again', () => {
  const tree = renderer.create(
    <Layout
      viewBox={{ x: -50, y: -50, width: 100, height: 100 }}
      className={'test2'}
      flat={false}
      origin={{ x: 0, y: 0 }}
      size={{ x: 12, y: 14 }}
      spacing={2.0}
    >
      {GridGenerator.hexagon(10).map(hex => <Hexagon {...hex} />)};
    </Layout>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

