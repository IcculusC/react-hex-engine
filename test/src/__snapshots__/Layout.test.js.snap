// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should only render hexagons in bounds 1`] = `
<g
  className="test2"
>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(0, 0)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
</g>
`;

exports[`Layout should only render hexagons in bounds again 1`] = `
<g
  className="test2"
>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(-41.569219381653056, 0)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(-20.784609690826528, 42)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(-20.784609690826528, -42)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(0, 0)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(20.784609690826528, 42)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(20.784609690826528, -42)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(41.569219381653056, 0)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="10.392304845413264,6.999999999999999 7.347880794884119e-16,14 -10.392304845413264,6.999999999999999 -10.392304845413266,-6.9999999999999964 -2.204364238465236e-15,-14 10.39230484541326,-7.000000000000006"
      />
    </g>
  </g>
  ;
</g>
`;

exports[`Layout should render correctly with custom props 1`] = `
<g
  className="test2"
>
  <div>
    child
  </div>
</g>
`;

exports[`Layout should render correctly with default props 1`] = `
<g
  className="test1"
>
  <div>
    child
  </div>
</g>
`;
